bundles/io/org.eclipse.smarthome.io.http/src/main/java/org/eclipse/smarthome/io/http/servlet/BaseSmartHomeServlet.java

```diff
@@ -0,0 +1,68 @@
+/**
+ * Copyright (c) 2014,2018 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.eclipse.smarthome.io.http.servlet;
+
+import java.util.Hashtable;
+
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServlet;
+
+import org.osgi.service.http.HttpContext;
+import org.osgi.service.http.HttpService;
+import org.osgi.service.http.NamespaceException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Base class for HTTP servlets.
+ *
+ * @author Łukasz Dywicki - Initial contribution and API
+ */
+public abstract class BaseSmartHomeServlet extends HttpServlet {
```

Comment:
Question: Instead of providing abstract base classes that need to be extended by everyone, wouldn't it be nicer to follow the whiteboard pattern and only have those implement a "SmartHomeServlet" interface, while the framework provides a service that collects those and registers them on the http service? From the past we have some bad experiences with using base classes as an API and thus would want to go for proper interfaces whereever possible. Wdyt?

Reply:
Difficulty with whiteboard is marrying everything together - you register things using `SmartHomeServlet` interface. In order to bind it with http service you need to make explicit cast to `HttpServlet`. It would be much better if we would have `SmartHomeServletProvider`, because then we avoid casts, however this interface would be a factory and updates of it will become harder.

We will still need to do binding of http context, the only difference is that we will do that centrally in whiteboard extender. Beyond that we will need to fetch servlet path (probably one of methods to be added in interface).
I understand your argumentation and motivation, however I wouldn't like to push that change over with this PR to do not mix things. It is definitely an improvement to above base class, however if we want to stick with this approach we need to decide what goes first - auth or whiteboard.

Reply:
Ok, let's not let the scope explode and leave the whiteboard refactoring for later. It just didn't look nice to have so many places in the code now extending a new abstract class although we do not want to use this pattern.